crates/driver/src/domain/competition/mod.rs

```diff
@@ -211,13 +213,20 @@ impl Competition {
             })
             .unzip();
 
-        self.settlement.lock().unwrap().clone_from(&settlement);
-
-        let settlement = match settlement {
-            Some(settlement) => settlement,
+        let Some(settlement) = settlement else {
             // Don't wait for the deadline because we can't produce a solution anyway.
-            None => return Ok(score),
+            return Ok(score);
         };
+        let solution_id = settlement.solution().get();
+
+        {
+            let mut lock = self.settlements.lock().unwrap();
+            lock.push_front(settlement.clone());
+
+            /// Number of solutions that may be cached at most.
+            const MAX_SOLUTION_STORAGE: usize = 5;
```

Comment:
There is no science behind me picking 5 here. The value should be big enough to allow some amount of successive wins but small enough to prevent solvers from submitting ancient solutions.
We could make this configurable but that seemed a bit overkill to me.